lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -2947,67 +2971,350 @@ bool InstCombiner::DoOneIteration(Function &F, unsigned Iteration) {
+
+// void LLVMInitializeInstCombine(LLVMPassRegistryRef R) {                                                                                                           
+//   initializeSeaInstCombine(*unwrap(R));                                                                                                                           
+// }     
+
+
+char llvm_seahorn::SeaInstructionCombiningPass::ID = 0;
+
+FunctionPass *llvm_seahorn::createInstructionCombiningPass() {
+  return new llvm_seahorn::SeaInstructionCombiningPass();
 }
 
-llvm::FunctionPass *llvm_seahorn::createInstructionCombiningPass() {
-  return new llvm_seahorn::InstCombiner();
+static llvm::RegisterPass<llvm_seahorn::SeaInstructionCombiningPass>
+X ("sea-instcombine", "SeaHorn instruction combiner");
+
+
+void InstructionCombiningPass::getAnalysisUsage(AnalysisUsage &AU) const {
+  AU.setPreservesCFG();
+  AU.addRequired<AAResultsWrapperPass>();
+  AU.addRequired<AssumptionCacheTracker>();
+  AU.addRequired<TargetLibraryInfoWrapperPass>();
+  AU.addRequired<DominatorTreeWrapperPass>();
+  AU.addPreserved<DominatorTreeWrapperPass>();
+  AU.addPreserved<AAResultsWrapperPass>();
+  AU.addPreserved<BasicAAWrapperPass>();
+  AU.addPreserved<GlobalsAAWrapperPass>();
+}
+
+bool InstructionCombiningPass::runOnFunction(Function &F) {
+  if (skipFunction(F))
+    return false;
+
+  // Required analyses.
+  auto AA = &getAnalysis<AAResultsWrapperPass>().getAAResults();
+  auto &AC = getAnalysis<AssumptionCacheTracker>().getAssumptionCache(F);
+  auto &TLI = getAnalysis<TargetLibraryInfoWrapperPass>().getTLI();
+  auto &DT = getAnalysis<DominatorTreeWrapperPass>().getDomTree();
+
+  // Optional analyses.
+  auto *LIWP = getAnalysisIfAvailable<LoopInfoWrapperPass>();
+  auto *LI = LIWP ? &LIWP->getLoopInfo() : nullptr;
+
+  return combineInstructionsOverFunction(F, Worklist, AA, AC, TLI, DT,
+                                         ExpensiveCombines, LI);
+}
+
+char InstructionCombiningPass::ID = 0;
+INITIALIZE_PASS_BEGIN(InstructionCombiningPass, "instcombine",
+                      "Combine redundant instructions", false, false)
+INITIALIZE_PASS_DEPENDENCY(AssumptionCacheTracker)
+INITIALIZE_PASS_DEPENDENCY(TargetLibraryInfoWrapperPass)
+INITIALIZE_PASS_DEPENDENCY(DominatorTreeWrapperPass)
+INITIALIZE_PASS_DEPENDENCY(AAResultsWrapperPass)
+INITIALIZE_PASS_DEPENDENCY(GlobalsAAWrapperPass)
+INITIALIZE_PASS_END(InstructionCombiningPass, "instcombine",
+                    "Combine redundant instructions", false, false)
+
+// Initialization Routines
+void llvm::initializeInstCombine(PassRegistry &Registry) {
+  initializeInstructionCombiningPassPass(Registry);
 }
 
-static llvm::RegisterPass<llvm_seahorn::InstCombiner>
-X ("sea-instcombine", "Instruction combiner");
+void LLVMInitializeInstCombine(LLVMPassRegistryRef R) {
+  initializeInstructionCombiningPassPass(*unwrap(R));
+}
 
+FunctionPass *llvm::createInstructionCombiningPass(bool ExpensiveCombines) {
```

Comment:
Why are we keeping `llvm::InstructionCombiningPass`? We just care about `llvm_seahorn::SeaInstructionCombiningPass`